File: android/shared/src/main/kotlin/app/paddleup/shared/gpx/GpxExporter.kt
package app.paddleup.shared.gpx

import app.paddleup.shared.model.Session
import app.paddleup.shared.model.TrackPoint
import kotlinx.datetime.Instant
import kotlin.time.Duration.Companion.seconds

/**
 * Generates a GPX 1.1 document for a [Session] + its full-resolution track.
 *
 * Outputs Garmin-compatible TrackPointExtension v2 elements for HR and cadence,
 * and a Paddleup-namespaced extension for speed and cadence-confidence.
 */
object GpxExporter {

    private const val GPXTPX = "http://www.garmin.com/xmlschemas/TrackPointExtension/v2"
    private const val PU = "http://paddleup.app/xmlschemas/v1"

    fun toGpx(session: Session, track: List<TrackPoint>): String {
        val started = Instant.parse(session.startedAt)
        val sb = StringBuilder(64 + track.size * 160)
        sb.append("""<?xml version="1.0" encoding="UTF-8"?>""").append('\n')
        sb.append(
            """<gpx version="1.1" creator="Paddleup"
              | xmlns="http://www.topografix.com/GPX/1/1"
              | xmlns:gpxtpx="$GPXTPX"
              | xmlns:pu="$PU">""".trimMargin().replace("\n", " "),
        ).append('\n')

        sb.append("  <metadata>\n")
        sb.append("    <name>").append(escape("Paddleup ${session.craftType.name} session")).append("</name>\n")
        sb.append("    <time>").append(session.startedAt).append("</time>\n")
        sb.append("  </metadata>\n")

        sb.append("  <trk>\n    <name>").append(session.id).append("</name>\n    <trkseg>\n")
        for (p in track) {
            val tIso = started.plus(p.t.seconds).toString()
            sb.append("      <trkpt lat=\"").append(p.lat).append("\" lon=\"").append(p.lon).append("\">\n")
            sb.append("        <ele>").append(p.altM).append("</ele>\n")
            sb.append("        <time>").append(tIso).append("</time>\n")
            sb.append("        <extensions>\n")
            if (p.hr != null || p.strokeRate != null) {
                sb.append("          <gpxtpx:TrackPointExtension>\n")
                p.hr?.let { sb.append("            <gpxtpx:hr>").append(it).append("</gpxtpx:hr>\n") }
                p.strokeRate?.let { sb.append("            <gpxtpx:cad>").append(it.toInt()).append("</gpxtpx:cad>\n") }
                sb.append("          </gpxtpx:TrackPointExtension>\n")
            }
            sb.append("          <pu:speed>").append(p.speedMps).append("</pu:speed>\n")
            p.cadenceConfidence?.let {
                sb.append("          <pu:cadConfidence>").append(it).append("</pu:cadConfidence>\n")
            }
            sb.append("        </extensions>\n")
            sb.append("      </trkpt>\n")
        }
        sb.append("    </trkseg>\n  </trk>\n</gpx>\n")
        return sb.toString()
    }

    private fun escape(s: String): String =
        s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
}

File: ios/PaddleupShared/Sources/PaddleupShared/GpxExporter.swift
import Foundation

public enum GpxExporter {

    private static let gpxtpx = "http://www.garmin.com/xmlschemas/TrackPointExtension/v2"
    private static let pu = "http://paddleup.app/xmlschemas/v1"

    public static func toGpx(session: Session, track: [TrackPoint]) -> String {
        let isoFormatter = ISO8601DateFormatter()
        isoFormatter.formatOptions = [.withInternetDateTime]
        let started = isoFormatter.date(from: session.startedAt) ?? Date()

        var s = ""
        s.reserveCapacity(64 + track.count * 160)
        s += #"<?xml version="1.0" encoding="UTF-8"?>"# + "\n"
        s += #"<gpx version="1.1" creator="Paddleup" xmlns="http://www.topografix.com/GPX/1/1" xmlns:gpxtpx="\#(gpxtpx)" xmlns:pu="\#(pu)">"# + "\n"
        s += "  <metadata>\n"
        s += "    <name>\(escape("Paddleup \(session.craftType.rawValue) session"))</name>\n"
        s += "    <time>\(session.startedAt)</time>\n"
        s += "  </metadata>\n"
        s += "  <trk>\n    <name>\(session.id)</name>\n    <trkseg>\n"

        for p in track {
            let date = started.addingTimeInterval(p.t)
            let tIso = isoFormatter.string(from: date)
            s += "      <trkpt lat=\"\(p.lat)\" lon=\"\(p.lon)\">\n"
            s += "        <ele>\(p.altM)</ele>\n"
            s += "        <time>\(tIso)</time>\n"
            s += "        <extensions>\n"
            if p.hr != nil || p.strokeRate != nil {
                s += "          <gpxtpx:TrackPointExtension>\n"
                if let hr = p.hr { s += "            <gpxtpx:hr>\(hr)</gpxtpx:hr>\n" }
                if let cad = p.strokeRate { s += "            <gpxtpx:cad>\(Int(cad))</gpxtpx:cad>\n" }
                s += "          </gpxtpx:TrackPointExtension>\n"
            }
            s += "          <pu:speed>\(p.speedMps)</pu:speed>\n"
            if let conf = p.cadenceConfidence {
                s += "          <pu:cadConfidence>\(conf)</pu:cadConfidence>\n"
            }
            s += "        </extensions>\n"
            s += "      </trkpt>\n"
        }
        s += "    </trkseg>\n  </trk>\n</gpx>\n"
        return s
    }

    private static func escape(_ s: String) -> String {
        s.replacingOccurrences(of: "&", with: "&amp;")
         .replacingOccurrences(of: "<", with: "&lt;")
         .replacingOccurrences(of: ">", with: "&gt;")
    }
}
